File: tools/multipart-archiver/multiparc.py
#!/usr/bin/python
# Multipart-"архиватор" для заливки многих изображений на сервер разом через импорт
# USAGE: ./multiparc.py file1 file2 file3 > outfile

import os, sys, time, hashlib
from stat import *
from xml.sax.saxutils import escape

if len(sys.argv) > 1:
    boundary = '--' + str(int(time.time()))
    sys.stdout.write("Content-Type: multipart/related; boundary="+boundary+"\n"+boundary+"\nContent-Type: text/xml\nContent-ID: Revisions\n\n");
    sys.stdout.write('<?xml version="1.0" ?>\n<mediawiki xmlns="http://www.mediawiki.org/xml/export-0.3/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mediawiki.org/xml/export-0.3/ http://www.mediawiki.org/xml/export-0.3.xsd" version="0.3" xml:lang="ru">\n')
    sizes = {}
    for arg in sys.argv[1:]:
        try:
            st = os.stat(arg)
            sizes[arg] = st[ST_SIZE]
            fp = open(arg, "rb")
            s = hashlib.sha1()
            s.update(fp.read(sizes[arg]))
            sha1 = s.hexdigest()
            t = arg
            p = t.rfind('/')
            if p >= 0:
                t = t.substr(p+1)
            t = t.replace(':', '-')
            sys.stdout.write(
'''<page>
 <title>Image:%s</title>
 <upload>
  <timestamp>%s</timestamp>
  <filename>%s</filename>
  <src sha1="%s">multipart://%s</src>
  <size>%d</size>
 </upload>
</page>\n''' % (escape(t), time.strftime('%Y-%m-%dT%H:%M:%SZ', time.localtime(st[ST_MTIME])), escape(arg), escape(sha1), escape(arg), sizes[arg]))
            fp.close()
        except:
            sys.stderr.write('Failed to add %s\n' % (arg))
    sys.stdout.write('</mediawiki>\n')
    for arg in sys.argv[1:]:
        try:
            fp = open(arg, "rb")
            sys.stdout.write(boundary+"\nContent-Type: application/binary\nContent-Transfer-Encoding: Little-Endian\nContent-ID: "+arg+"\nContent-Length: "+str(sizes[arg])+"\n\n")
            sys.stdout.write(fp.read(sizes[arg]))
            fp.close()
        except:
            sys.stderr.write('Failed to add %s\n' % (arg))
else:
    sys.stderr.write("USAGE: python multiparc.py FILE1 FILE2 FILE3 > OUTFILE\n")

File: tools/multipart-archiver/multiparc.php
#!/usr/bin/php
<?php
# Multipart-"архиватор" для заливки на сервер множества изображений разом через импорт
# USAGE: ./multiparc.py file1 file2 file3 > outfile

if (count($argv) == 1)
{
    print "USAGE: php multiparc.php FILE1 FILE2 FILE3 > OUTFILE\n";
    exit;
}

$boundary = '--' . time();
print "Content-Type: multipart/related; boundary=$boundary
$boundary
Content-Type: text/xml
Content-ID: Revisions

".'<?xml version="1.0" encoding="UTF-8" ?>
<mediawiki xmlns="http://www.mediawiki.org/xml/export-0.3/"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://www.mediawiki.org/xml/export-0.3/ http://www.mediawiki.org/xml/export-0.3.xsd"
    version="0.3" xml:lang="ru">
';

$sizes = array();
for ($i = 1; $i < count($argv); $i++)
{
    $filename = $argv[$i];
    $stat = stat($filename);
    if (!$stat)
    {
        print "File not found: $filename\n";
        continue;
    }
    $sha1 = sha1_file($filename);
    $wikiname = htmlspecialchars(str_replace(':', '-', preg_replace('#^.*/#is', '', $filename)));
    $files[$wikiname] = $filename;
    $timestamp = strftime('%Y-%m-%dT%H:%M:%SZ', $stat['mtime']);
    print "<page>
 <title>File:$wikiname</title>
 <upload>
  <timestamp>$timestamp</timestamp>
  <filename>$wikiname</filename>
  <src sha1=\"$sha1\">multipart://$wikiname</src>
  <size>$stat[size]</size>
 </upload>
</page>";
}
print "</mediawiki>\n";

foreach ($files as $wikiname => $filename)
{
    $data = file_get_contents($filename);
    print "$boundary
Content-Type: application/binary
Content-Transfer-Encoding: Little-Endian
Content-ID: $wikiname
Content-Length: ".strlen($data)."

$data";
}
